### PR TITLE
i#5520 memtrace encodings: Add AArch32 support

### PR DESCRIPTION
**Pull request by derekbruening**
Adds support for Thumb versus Arm modes in AArch32 with respect to drmemtrace stored encodings by setting the LSB when storing in the tracer and raw2trace, and by properly setting the base mode to Arm when decoding.

Augments decode_from_copy() to switch locally to Thumb if either the read or original PC has LSB=1 as that better fits most use cases (including reading from a drmemtrace encoding buffer).

Adds AArch32 support to the burst_gencode test.

Fixes 3 outstanding AArch32 issues:

+ Removes a quote in a comment in third_party/libgcc/arm/lib1funcs.S code not handled by old toolchains.

+ Removes an ASSERT_NOT_TESTED path now hit.

+ Fixes an AArch32 tracer bug where a 2nd temp register is needed for a 2nd conditional skip.

Tested by running the drcacheoff.gencode test on an AArch32 machine which now passes.

Issue: #5520

### CODE REVIEW

File: third_party/libgcc/arm/lib1funcs.S [hunk 1]
```diff
@@ -1465,7 +1465,8 @@ LSYM(Lover12):
         asr     ah, r2
         sub     r2, #32
         @ If r2 is negative at this point the following step would OR
-        @ the sign bit into all of AL.  That's not what we want...
+        @ <bruening: removing quote on next line since one toolchain trips over it>
+        @ the sign bit into all of AL.  Thats not what we want...
```
Comment (abhinav92003) [inline third_party/libgcc/arm/lib1funcs.S:1469]: s/Thats/That is/
or
s/Thats not what we want/We do not want that/

Is the comment in the above line intended to be kept?

**Review by abhinav92003 — APPROVED** [review]

File: third_party/libgcc/arm/lib1funcs.S (hunk 1 above)
Comment (derekbruening) [reply third_party/libgcc/arm/lib1funcs.S:1469]: > Is the comment in the above line intended to be kept?

Yes, so future devs aren't confused by the diff vs upstream lib1funcs.S.  Looks like I should mention this or include the diff in ../README.dynamorio too: will do that.

Comment (derekbruening) [reply third_party/libgcc/arm/lib1funcs.S:1469]: Maybe just the mention in the README is enough and we don't need the embedded comment.

**Comment by derekbruening** [comment]
32-bit failure is scattergather pipe assert #5329

**Pull request merged**